### PR TITLE
[Fluid][SDEM] Redefine manufactured solution

### PR DESCRIPTION
**Pull request by jgonzalezusua**
**Description**
There are some objectives of this PR:

- Include the Drew's formulation for flows in porous media using the VMS framework.
- Implement Dallavalle's drag law to simulate dense-packed beds.
- Including more manufactured solutions: hyperbolic tangential and sinusoidal.
- Implement a new diffusion-based coarse graining function.

Please mark the PR with appropriate tags: 
- Applications
- Feature

**Changelog**
The changes made in this PR are the following:

- Implement two new elements using the Drew's formulation: `alternative_qs_vms_dem_coupled.cpp` and  `alternative_d_vms_dem_coupled.cpp`.
- Modify `bdf2_turbulent_schemeDEMCoupled.h` to update the residual projection before calculation and the subscale after it.
- Including a bool to project the body force depending on the type of model used (Drew or Jackson) in `binbased_DEM_fluid_coupled_mapping.cpp`.
- Changing the coarse graining function in `density_function_polynomial.h`, now its integral is equal to 1 and its derivatives in the boundaries are 0.
- Call to the particle instead of the node in the drag laws.
- Removing computation of the manufactured solution at each time step when they are not time dependant.

### CODE REVIEW

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
This PR is too large to be reviewed. I'd suggest to, at least, separate the `FluidDynamicsApplication` changes from the `DEMApplication` ones.

**Comment by jgonzalezusua** [comment]
> This PR is too large to be reviewed. I'd suggest to, at least, separate the `FluidDynamicsApplication` changes from the `DEMApplication` ones.

It is true that the PR is large, but as a fluid maintainer you can focus only on the FluidDynamicsApplications, most of the changes are related to the implementation of two new elements. The rest of the files can be reviewed by the DEM team.

**Comment by rubenzorrilla** [comment]
> > This PR is too large to be reviewed. I'd suggest to, at least, separate the `FluidDynamicsApplication` changes from the `DEMApplication` ones.
> 
> It is true that the PR is large, but as a fluid maintainer you can focus only on the FluidDynamicsApplications, most of the changes are related to the implementation of two new elements. The rest of the files can be reviewed by the DEM team.

I insist. In case something gets broken it would be very difficult and complex to backtrack the changes and eventually revert. Considering that it is straightforward to port the changes to from this branch to a new one (create a new branch, switch to it and do `git checkout Redefine-manufactured-solution my_fluid_file.cpp` for the files in the `FluidDynamicsApplication`) I'd appreciate doing the merge in such way. Then you just need to merge the master in here once the auxiliary PR is finished.

**Comment by jgonzalezusua** [comment]
> > > This PR is too large to be reviewed. I'd suggest to, at least, separate the `FluidDynamicsApplication` changes from the `DEMApplication` ones.
> > 
> > 
> > It is true that the PR is large, but as a fluid maintainer you can focus only on the FluidDynamicsApplications, most of the changes are related to the implementation of two new elements. The rest of the files can be reviewed by the DEM team.
> 
> I insist. In case something gets broken it would be very difficult and complex to backtrack the changes and eventually revert. Considering that it is straightforward to port the changes to from this branch to a new one (create a new branch, switch to it and do `git checkout Redefine-manufactured-solution my_fluid_file.cpp` for the files in the `FluidDynamicsApplication`) I'd appreciate doing the merge in such way. Then you just need to merge the master in here once the auxiliary PR is finished.

The key problem is that I am working in the SwimmingDEMApplication which derives in part from the FluidDynamicsApplication. I need to merge the changes in the fluid application together with the rest of the PR.

**Comment by rubenzorrilla** [comment]
> > > > This PR is too large to be reviewed. I'd suggest to, at least, separate the `FluidDynamicsApplication` changes from the `DEMApplication` ones.
> > > 
> > > 
> > > It is true that the PR is large, but as a fluid maintainer you can focus only on the FluidDynamicsApplications, most of the changes are related to the implementation of two new elements. The rest of the files can be reviewed by the DEM team.
> > 
> > 
> > I insist. In case something gets broken it would be very difficult and complex to backtrack the changes and eventually revert. Considering that it is straightforward to port the changes to from this branch to a new one (create a new branch, switch to it and do `git checkout Redefine-manufactured-solution my_fluid_file.cpp` for the files in the `FluidDynamicsApplication`) I'd appreciate doing the merge in such way. Then you just need to merge the master in here once the auxiliary PR is finished.
> 
> The key problem is that I am working in the SwimmingDEMApplication which derives in part from the FluidDynamicsApplication. I need to merge the changes in the fluid application together with the rest of the PR.

mmm... it is strange. The changes in the `FluidDynamicsApplication` must compile without the `SwimmingDEMApplication`. Otherwise there would be a dependency in the other way around.

File: applications/FluidDynamicsApplication/custom_elements/fluid_element.h
Comment (philbucher) [inline applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: Please don’t add this, it is superfluous and clang format warns abt this
See https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-override.html

**Comment by jgonzalezusua** [comment]
> > > > > This PR is too large to be reviewed. I'd suggest to, at least, separate the `FluidDynamicsApplication` changes from the `DEMApplication` ones.
> > > > 
> > > > 
> > > > It is true that the PR is large, but as a fluid maintainer you can focus only on the FluidDynamicsApplications, most of the changes are related to the implementation of two new elements. The rest of the files can be reviewed by the DEM team.
> > > 
> > > 
> > > I insist. In case something gets broken it would be very difficult and complex to backtrack the changes and eventually revert. Considering that it is straightforward to port the changes to from this branch to a new one (create a new branch, switch to it and do `git checkout Redefine-manufactured-solution my_fluid_file.cpp` for the files in the `FluidDynamicsApplication`) I'd appreciate doing the merge in such way. Then you just need to merge the master in here once the auxiliary PR is finished.
> > 
> > 
> > The key problem is that I am working in the SwimmingDEMApplication which derives in part from the FluidDynamicsApplication. I need to merge the changes in the fluid application together with the rest of the PR.
> 
> mmm... it is strange. The changes in the `FluidDynamicsApplication` must compile without the `SwimmingDEMApplication`. Otherwise there would be a dependency in the other way around.

Which is true, the problem is in the SDEM application. Some tests need the new elements.

**Comment by rubenzorrilla** [comment]
> > > > > > This PR is too large to be reviewed. I'd suggest to, at least, separate the `FluidDynamicsApplication` changes from the `DEMApplication` ones.
> > > > > 
> > > > > 
> > > > > It is true that the PR is large, but as a fluid maintainer you can focus only on the FluidDynamicsApplications, most of the changes are related to the implementation of two new elements. The rest of the files can be reviewed by the DEM team.
> > > > 
> > > > 
> > > > I insist. In case something gets broken it would be very difficult and complex to backtrack the changes and eventually revert. Considering that it is straightforward to port the changes to from this branch to a new one (create a new branch, switch to it and do `git checkout Redefine-manufactured-solution my_fluid_file.cpp` for the files in the `FluidDynamicsApplication`) I'd appreciate doing the merge in such way. Then you just need to merge the master in here once the auxiliary PR is finished.
> > > 
> > > 
> > > The key problem is that I am working in the SwimmingDEMApplication which derives in part from the FluidDynamicsApplication. I need to merge the changes in the fluid application together with the rest of the PR.
> > 
> > 
> > mmm... it is strange. The changes in the `FluidDynamicsApplication` must compile without the `SwimmingDEMApplication`. Otherwise there would be a dependency in the other way around.
> 
> Which is true, the problem is in the SDEM application. Some tests need the new elements.

But that's precisely my point.
1. We create an auxiliary PR with the `FluidDynamicsApplication` changes. This must contain the at least a cpp test for the new elements. Otherwise it is impossible to check if something gets broken without compiling the `SwimmingDEMApplication`.
2. We review & merge the PR in 1
3. Merging the `master` in `Redefine-manufactured-solution` branch
4. We review & merge this PR.

**Comment by jgonzalezusua** [comment]
> > > > > > > This PR is too large to be reviewed. I'd suggest to, at least, separate the `FluidDynamicsApplication` changes from the `DEMApplication` ones.
> > > > > > 
> > > > > > 
> > > > > > It is true that the PR is large, but as a fluid maintainer you can focus only on the FluidDynamicsApplications, most of the changes are related to the implementation of two new elements. The rest of the files can be reviewed by the DEM team.
> > > > > 
> > > > > 
> > > > > I insist. In case something gets broken it would be very difficult and complex to backtrack the changes and eventually revert. Considering that it is straightforward to port the changes to from this branch to a new one (create a new branch, switch to it and do `git checkout Redefine-manufactured-solution my_fluid_file.cpp` for the files in the `FluidDynamicsApplication`) I'd appreciate doing the merge in such way. Then you just need to merge the master in here once the auxiliary PR is finished.
> > > > 
> > > > 
> > > > The key problem is that I am working in the SwimmingDEMApplication which derives in part from the FluidDynamicsApplication. I need to merge the changes in the fluid application together with the rest of the PR.
> > > 
> > > 
> > > mmm... it is strange. The changes in the `FluidDynamicsApplication` must compile without the `SwimmingDEMApplication`. Otherwise there would be a dependency in the other way around.
> > 
> > 
> > Which is true, the problem is in the SDEM application. Some tests need the new elements.
> 
> But that's precisely my point.
> 
> 1. We create an auxiliary PR with the `FluidDynamicsApplication` changes. This must contain the at least a cpp test for the new elements. Otherwise it is impossible to check if something gets broken without compiling the `SwimmingDEMApplication`.
> 2. We review & merge the PR in 1
> 3. Merging the `master` in `Redefine-manufactured-solution` branch
> 4. We review & merge this PR.

Ok, understood. I will do that.

**Review by philbucher — COMMENTED** [review]
There are many more of these virtual, I pointed it out only in one place

**Comment by jgonzalezusua** [comment]
> There are many more of these virtual, I pointed it out only in one place

Ok, dont worry, I will work on that

**Comment by rubenzorrilla** [comment]
> > > > > > > > This PR is too large to be reviewed. I'd suggest to, at least, separate the `FluidDynamicsApplication` changes from the `DEMApplication` ones.
> > > > > > > 
> > > > > > > 
> > > > > > > It is true that the PR is large, but as a fluid maintainer you can focus only on the FluidDynamicsApplications, most of the changes are related to the implementation of two new elements. The rest of the files can be reviewed by the DEM team.
> > > > > > 
> > > > > > 
> > > > > > I insist. In case something gets broken it would be very difficult and complex to backtrack the changes and eventually revert. Considering that it is straightforward to port the changes to from this branch to a new one (create a new branch, switch to it and do `git checkout Redefine-manufactured-solution my_fluid_file.cpp` for the files in the `FluidDynamicsApplication`) I'd appreciate doing the merge in such way. Then you just need to merge the master in here once the auxiliary PR is finished.
> > > > > 
> > > > > 
> > > > > The key problem is that I am working in the SwimmingDEMApplication which derives in part from the FluidDynamicsApplication. I need to merge the changes in the fluid application together with the rest of the PR.
> > > > 
> > > > 
> > > > mmm... it is strange. The changes in the `FluidDynamicsApplication` must compile without the `SwimmingDEMApplication`. Otherwise there would be a dependency in the other way around.
> > > 
> > > 
> > > Which is true, the problem is in the SDEM application. Some tests need the new elements.
> > 
> > 
> > But that's precisely my point.
> > 
> > 1. We create an auxiliary PR with the `FluidDynamicsApplication` changes. This must contain the at least a cpp test for the new elements. Otherwise it is impossible to check if something gets broken without compiling the `SwimmingDEMApplication`.
> > 2. We review & merge the PR in 1
> > 3. Merging the `master` in `Redefine-manufactured-solution` branch
> > 4. We review & merge this PR.
> 
> Ok, understood. I will do that.

Cool. Just let me know if you have any issue. Thanks!!! :rocket:

**Review by rubenzorrilla — DISMISSED** [review]
No more concerns from the fluid side. I let the @KratosMultiphysics/dem guys to give the final approval.

**Review by maceligueta — APPROVED** [review]
No problems from my side

File: applications/SwimmingDEMApplication/custom_constitutive/drag_laws/chien_drag_law.cpp
Comment (maceligueta) [inline applications/SwimmingDEMApplication/custom_constitutive/drag_laws/chien_drag_law.cpp]: reference here?

**Review by maceligueta — CHANGES_REQUESTED** [review]

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp
Comment (maceligueta) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp]: why this alias? you could use mViscosity directly everywhere...

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp
Comment (maceligueta) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp]: why this alias?

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp
Comment (maceligueta) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp]: Why alias?

File: applications/SwimmingDEMApplication/custom_processes/transient_spatial_dependant_porosity_solution_body_force_process.cpp
Comment (maceligueta) [inline applications/SwimmingDEMApplication/custom_processes/transient_spatial_dependant_porosity_solution_body_force_process.cpp]: Alias (reference)

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp
Comment (jgonzalezusua) [reply applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp]: Leftlover, I forgot to remove it

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_hyperbolic_tangential_porosity_solution_and_body_force_process.cpp
Comment (maceligueta) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_hyperbolic_tangential_porosity_solution_and_body_force_process.cpp]: here too

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_transient_body_force_process.cpp
Comment (maceligueta) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_transient_body_force_process.cpp]: These two could be a const reference instead of a copy, your call

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_transient_body_force_process.cpp
Comment (maceligueta) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_transient_body_force_process.cpp]: const if possible, not sure if you are changing these values

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_transient_body_force_process.cpp
Comment (maceligueta) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_transient_body_force_process.cpp]: ref?

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_sinusoidal_porosity_solution_and_body_force_process.cpp
Comment (maceligueta) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_sinusoidal_porosity_solution_and_body_force_process.cpp]: ref maybe

**Review by maceligueta — DISMISSED** [review]
I am approving because my proposed changes do not affect performance, I think. It is your call to modify those lines and similar, or not.

**Review by maceligueta — DISMISSED** [review]

File: applications/SwimmingDEMApplication/custom_constitutive/drag_laws/dallavalle_drag_law.cpp
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_constitutive/drag_laws/dallavalle_drag_law.cpp:2]: Please, add a short comment metnioning the origin of the formulation (academic paper) along with any changes or errata corrections with respect to the original source.

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_hyperbolic_tangential_porosity_solution_and_body_force_process.cpp
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_hyperbolic_tangential_porosity_solution_and_body_force_process.cpp]: Please, change the name of the process to something more exactly descriptive. Example: hyperbolic_tangent_porosity_solution_and_body_force_process.cpp
(I would skip the 'spatially_dependant' prefix, since it can be assumed to be virtually always)

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp]: Please add comments identifying the python code with which the hard-coded formulas where derived.

Comment (jgonzalezusua) [reply applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_and_body_force_process.cpp]: The python code is external from Kratos

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_sinusoidal_porosity_solution_and_body_force_process.cpp
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_sinusoidal_porosity_solution_and_body_force_process.cpp]: same comment about referring the python code from which the formulas are derived

File: applications/SwimmingDEMApplication/custom_utilities/binbased_DEM_fluid_coupled_mapping.cpp
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/binbased_DEM_fluid_coupled_mapping.cpp:227]: all these parameters should me moved to a Parameters object. For the future.

File: applications/SwimmingDEMApplication/custom_utilities/custom_functions.h
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/custom_functions.h]: I would prefer to leave it imprecise: it is not going to be exactly zero. Better to normalize the volume first by the size of the element cubed and then check with the epsilon. That is,
`if (std::abs(vol)/std::pow(h, 3) < std::numeric_limits<double>::epsilon())`
where
`const double h = CalculateDiameter(x0, y0, z0, x1, y1, z1, x2, y2, z2, x3, y3, z3)`
And where the diameter is easily computed as the maximum of the distances between pairs of points of the geometry.

File: applications/SwimmingDEMApplication/custom_utilities/mollification/density_function_polynomial.h
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/mollification/density_function_polynomial.h]: use std::pow now that you are add it

File: applications/SwimmingDEMApplication/python_scripts/apply_spatial_dependant_hyperbolic_tangential_porosity_solution_and_body_force_process.py
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/python_scripts/apply_spatial_dependant_hyperbolic_tangential_porosity_solution_and_body_force_process.py]: the name of this process will be affected by the name changes above... There should be a better way, where you apply any 'solution_and_force' process from which the others derive. This can be left for the future too.

File: applications/SwimmingDEMApplication/python_scripts/fluid_fraction_test_analysis.py
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/python_scripts/fluid_fraction_test_analysis.py]: too long, take it to the next line

File: applications/SwimmingDEMApplication/python_scripts/hdf5_script.py
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/python_scripts/hdf5_script.py:99]: This is too slow.. Please, predimension numpy arrays for this, do not append

File: applications/SwimmingDEMApplication/python_scripts/hdf5_script.py
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/python_scripts/hdf5_script.py]: too long: use multiple lines

File: applications/SwimmingDEMApplication/python_scripts/hdf5_script.py
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/python_scripts/hdf5_script.py]: idem

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py]: too long: take to next line

File: applications/SwimmingDEMApplication/python_scripts/variables_management.py
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/python_scripts/variables_management.py]: better use 'FLUID_FRACTION_OLD_2' or 'FLUID_FRACTION_OLD_OLD'

File: applications/SwimmingDEMApplication/tests/PorosityTestFactory.py
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/tests/PorosityTestFactory.py:7]: add brief explanation of what the test does

File: applications/SwimmingDEMApplication/python_scripts/hdf5_script.py
Comment (jgonzalezusua) [reply applications/SwimmingDEMApplication/python_scripts/hdf5_script.py:99]: This process was created to validate new drag force laws such as the Dallavalle's one. I think the performance of the `append` it is not important.

Comment (GuillermoCasas) [reply applications/SwimmingDEMApplication/python_scripts/hdf5_script.py:99]: Ok, but you must create a test for this Drag model in the future. When you do, please eliminate or modify this function.

File: applications/SwimmingDEMApplication/custom_constitutive/drag_laws/dallavalle_drag_law.cpp
Comment (GuillermoCasas) [reply applications/SwimmingDEMApplication/custom_constitutive/drag_laws/dallavalle_drag_law.cpp:2]: Please add author and year and check spelling

**Review by GuillermoCasas — DISMISSED** [review]
Good job

**Review by GuillermoCasas — DISMISSED** [review]

**Review by ipouplana — APPROVED** [review]

**Pull request merged**
